napari/viewer.py

```diff
@@ -79,8 +82,26 @@ def screenshot(self, with_viewer=False):
         """
         if with_viewer:
             image = self.window.screenshot()
+
         else:
+            # capture initial canvas size to reset after screenshot
```

Comment:
I think here I'd rather take the screenshot as is and then force the resizing using the [skimage resize function](https://scikit-image.org/docs/stable/api/skimage.transform.html#skimage.transform.resize). This should avoid triggering the canvas to actually change size and should be much faster (I guess).

Using `resize` will also allow us to solve our aspect ratio problem as it allows you to pass a tuple for shape and specify things like zero padding using the `mode` keyword argument. We should then replace the `max_shape` keyword argument with one just called `shape` which is a 2-tuple (which you may need to add a [3] onto as the screenshots give back rgb images I think)

This approach should now guarantee that if we request say a `[300, 400]` image we'll always get one and should work regardless of using `with_viewer` or not as it only operates on the image after the screenshot.

If you can get this working can you also add some tests to `napari/tests/test_screenshot.py` which test both `with_viewer` and `shape`, and maybe move our existing screenshot test in there too - https://github.com/napari/napari/blob/fffd2f2a4090f773f2aec904bee9ac28a4b1890d/napari/tests/test_viewer.py#L79.

Reply:
I'm not convinced that this is a good solution. Resizing the images will deteriorate the output. Increasing the canvas size effectively captures a larger image, while ```skimage.transfrom.resize()``` just interpolates. As an example I add here two images which in the end have the same size, but one is resized with skimage (top) and for the other the canvas size is adjusted (bottom). 
![with_skimage](https://user-images.githubusercontent.com/4622767/70978159-891d8400-20af-11ea-8170-33ca199c428e.png)
![with_canvas](https://user-images.githubusercontent.com/4622767/70978174-8fabfb80-20af-11ea-93b4-4abbc8926f53.png)

Maybe @jni has an opinion here ?

Reply:
I don't, sorry! 😂 @sofroniewn is speed the only consideration with canvas resizing? Or is it also because the viewer itself will now change size? I agree that the latter is not ideal. But, the image quality problem is significant. Above, it's subtle, but visible. In other cases it might make a major difference.

Reply:
I just want to add here that the whole function is written in such a way that the canvas is automatically reset after movie making, so this is transparent to the user. So the only issue might be speed. I didn't do extensive tests, but it never seemed to be an issue to me.

Reply:
@guiwitz your point about the high quality rendering is well taken, but if people don't provide a scale-like argument here they will still get back the full quality stuff and I imagine that if they are providing a scale-like argument they might be doing so to get something much smaller anyway. I will add though that @royerloic and @jni did seem to prefer the high quality renderings though, so I don't think this is settled yet.

I've also only had time to test the functionality in a cursory way myself, but I am noting some unexpected behaviors when the rescaling is done this way, potentially caused by limits Qt places on how it does the rescaling - for example if I run `examples/add_image.py` to display the standard 2D astronaut picture and then run commands like the following from the console I get
```python
viewer.screenshot(max_shape=None).shape # (796, 1600, 4)
viewer.screenshot(max_shape=1000).shape # (994, 2000, 4)
viewer.screenshot(max_shape=100).shape # (400, 400, 4)
viewer.screenshot(max_shape=10).shape # (400, 400, 4)
```
whereas i would have expected to get
```python
viewer.screenshot(max_shape=None).shape # (796, 1600, 4)
viewer.screenshot(max_shape=1000).shape # (498, 1000, 4)
viewer.screenshot(max_shape=100).shape # (50, 100, 4)
viewer.screenshot(max_shape=10).shape # (5, 10, 4)
```

I still lean towards using `skimage.transfrom.resize` but maybe figuring out how to make sure people don't call it when unnecessarily - when they want the full resolution stuff, and to make sure it does the best things possible when it is doing downsampling etc.

I'll try and have a more detailed look at this again soon, but will be traveling for the holidays now

Reply:
following up, it seems like there is a factor of 2 problem for intermediate sizes, small sizes are limited by our Qt limits, and big sizes (say >20_000) cause vispy crashed. if there is a way to get the behaviour working with `skimage.transfrom.resize` as best as possible I think it would be worth trying